pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -277,3 +294,44 @@ func (c *DeploymentConfigController) updateStatus(config *deployapi.DeploymentCo
 	glog.V(4).Infof("Updated the status for %q (observed generation: %d)", deployutil.LabelForDeploymentConfig(config), config.Status.ObservedGeneration)
 	return nil
 }
+
+// cleanupOldDeployments triggers on deployment configuration update and add events in order
+// to delete old replication controller deployments if their quota has been reached
+func (c *DeploymentConfigController) cleanupOldDeployments(existingDeploymentList *kapi.ReplicationControllerList, deploymentConfig *deployapi.DeploymentConfig) error {
+	activeDeployment := deployutil.ActiveDeployment(deploymentConfig, existingDeployments)
+
+	// we need to coerce the list type into something we can sort, and we need to filter out
+	// any deployments that we don't care about, namely the active deployment and any newer
+	// deployments
+	existingDeployments := []*kapi.ReplicationController{}
+	for i := 0; i < len(existingDeploymentList.Items); i++ {
+		if existingDeploymentList.Items[i] == activeDeployment || DeploymentVersionFor(existingDeploymentList.Items[i]) > DeploymentVersionFor(activeDeployment) {
+			continue
+		}
+		existingDeployments = append(existingDeployments, &existingDeploymentList.Items[i])
+	}
+
+	if deploymentConfig.Spec.RevisionHistoryLimit == nil || len(existingDeployments) <= *deploymentConfig.Spec.RevisionHistoryLimit {
+		// the past deployment quota has either not been set or has not been exceeded
+		return nil
+	}
+
+	sort.Sort(controller.ControllersByCreationTimestamp(existingDeployments))
+
+	deletionErrors := []error{}
+	for i := 0; i < (len(existingDeployments) - *deploymentConfig.Spec.RevisionHistoryLimit); i++ {
+		deployment := existingDeployments[i]
+		// avoid cleaning up synced deployments or those with active pods
+		if deployment.Spec.Replicas != 0 && !deployutil.HasSynced(deploymentConfig) {
```

Comment:
HasSynced is supposed to be called by clients and it lets them know that the controller has observed the latest state of the deploymentconfig. It makes no sense to call it here, it will help you in your extended test.